### PR TITLE
Move post compile steps into post_compile1/post_compile2 method

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141695
* #141691
* #141689
* #141688
* #141685
* #141683
* #141681
* __->__ #141656

The intention for turning these into methods is so that the AOTInductor compile path can implement them differently. I haven't worked out the implications yet though, but this seemed like a good stopping point for now.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/output_code.py [hunk 1]
```diff
@@ -154,6 +222,96 @@ def __call__(self, inputs: Sequence[Any]) -> Any:
...
 
+    def post_compile1(
+        self,
+        cudagraphs: BoxedBool,
+        example_inputs: Sequence[InputType],
+        gm: GraphModule,
+        static_input_idxs: Sequence[int],
+        fx_kwargs: _CompileFxKwargs,
+        inputs_to_check: Sequence[int],
+        boxed_forward_device_index: Optional[BoxedDeviceIndex],
```
Comment (ezyang) [inline torch/_inductor/output_code.py:233]: I just bashed arguments in here until all the free variables were lifted as arguments. It probably needs a little reordering / coalescing, will figure this out on the next PR.

File: torch/_inductor/compile_fx.py [hunk 2]
```diff
@@ -771,9 +669,7 @@ def codegen_and_compile(
                 # In that case, we don't need to run all post compilation steps, we just need
                 # to return the string directly.
                 return compiled_graph
-            compiled_graph = FxGraphCache.post_compile(
-                compiled_graph, example_inputs, cudagraphs, gm
-            )
+            compiled_graph.post_compile2(example_inputs, cudagraphs, gm)
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:672]: Why is there post_compile1/post_compile2? It has to do with the FxGraphCache.load path above. I plan to refactor that too, but this is an intermediate step.

File: torch/_inductor/compile_fx.py [hunk 3]
```diff
@@ -664,78 +625,15 @@ def codegen_and_compile(
                 # need to do any post-compilation steps: we just return the string,
                 # which is the filename of the compiled code.
                 return compiled_graph
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:627]: You can see above that post compile does nothing in AOT mode. I want to eventually have AOTInductor do its own custom post compile logic at this point. However, I think there really should only be one post compile method.

File: torch/_inductor/compile_fx.py [hunk 4]
```diff
@@ -665,78 +629,15 @@ def codegen_and_compile(
...
-                    cudagraph_info = CudagraphCachedInfo(
-                        placeholders, stack_traces, cudagraph_fail_reasons
-                    )
-
-            compiled_graph.cudagraph_info = cudagraph_info
-            compiled_graph.inputs_to_check = inputs_to_check
-            compiled_graph.fx_kwargs = fx_kwargs
-            # TODO: should this be part of fx_kwargs
-            compiled_graph.boxed_forward_device_index = boxed_forward_device_index
+            compiled_graph.post_compile1(
```
Comment (jamesjwu) [inline torch/_inductor/compile_fx.py:632]: The difference between post_compile1 and post_compile2 is that post_compile1 happens pre serialization, and post_compile2 happens post serialization. They're both "post_compile", but I would honestly consider post_compile1 to really be "part of compile", if we believe that CompiledFxGraph is the final "compiled" product we get.

File: torch/_inductor/output_code.py [hunk 5]
```diff
@@ -55,6 +65,29 @@ class OutputCode(Protocol):
     def __call__(self, inputs: Sequence[Any]) -> Any:
         ...
 
+    # TODO: Not sure why we need two post_compile functions.  They're called
+    # in this order
+
+    def post_compile1(
```
Comment (jamesjwu) [inline torch/_inductor/output_code.py:71]: Maybe we call this one "pre serialize" or something

**Review by aorenste — APPROVED** [review]

**Review by jamesjwu — APPROVED** [review]
I would like to see a bigger picture set of tests at the end of the stack to make sure we hit all of the complicated code paths.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/compile_fx.py (hunk 4 above)
Comment (ezyang) [reply torch/_inductor/compile_fx.py:632]: post_compile1 doesn't actually happen pre serialization.  In https://github.com/pytorch/pytorch/pull/141689 we discover they are actually happening one after another and can be merged.

File: torch/_inductor/output_code.py [hunk 6]
```diff
@@ -55,6 +65,29 @@ class OutputCode(Protocol):
...
+        inputs_to_check: Sequence[int],
+        boxed_forward_device_index: Optional[BoxedDeviceIndex],
+    ) -> None:
+        ...
+
+    def post_compile2(
+        self,
+        example_inputs: Sequence[InputType],
+        cudagraphs: BoxedBool,
+        gm: GraphModule,
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_inductor/output_code.py:87]: Thanks for letting me know, I did not figure this out from my refactoring.

For non-AOTInductor cache path, I am interested in reducing the number of arguments that `post_compile` needs to take. But I am not too sure what actually can be simplified. I will need to attempt it to see.

**Pull request closed without merging**